### PR TITLE
XCOM-309: Added test coverage to the Course Admin Tool's create/edit view

### PR DESCRIPTION
**Pull request by Nickersoft**
@clintonb @rlucioni @jimabramson

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "backbone-route-filter": "~0.1.2",
     "backbone-relational": "~0.9.0",
     "backbone-validation": "~0.11.5",
-    "backbone.stickit": "~0.9.2"
+    "backbone.stickit": "~0.9.2",
+    "jquery-ui": "~1.11.4"
```
Comment (clintonb) [inline bower.json]: Remove. This isn't being used, and I have a general dislike for jQuery UI.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+              CourseCreateEditView,
+              Course) {
+        'use strict';
+
+        describe('course create view', function () {
+            var view,
+                model;
+
+            beforeEach(function () {
+                model = new Course({});
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: No need for the `{}`.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                var mockMouseHandler = { preventDefault: function () {} };
+                view.formView.submit(mockMouseHandler);
+                setTimeout(function () {
+                    expect(view.$el.find('.alert').length).toBeGreaterThan(0);
+                    done();
+                }, 100);
+            });
+        });
+
+        describe('course edit view', function () {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Move this to its own file, `course_edit_view_spec.js`

File: bower.json (hunk 1 above)
Comment (Nickersoft) [inline bower.json]: @clintonb .tooltip() is a jQuery-UI function:

https://github.com/edx/ecommerce/blob/dc318a57d8be24531753e2e0266a6e63862341d3/ecommerce/static/js/views/course_create_edit_view.js?code=373a654e587730e9d225#L54

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                var mockMouseHandler = { preventDefault: function () {} };
+                view.formView.submit(mockMouseHandler);
+                setTimeout(function () {
+                    expect(view.$el.find('.alert').length).toBeGreaterThan(0);
+                    done();
+                }, 100);
+            });
+
+            it('should throw no errors if ', function (done) {
+                var mockMouseHandler = { preventDefault: function () {} };
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Is this really necessary? Why not click the actual submit button?

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                view.formView.submit(mockMouseHandler);
+                setTimeout(function () {
+                    expect(view.$el.find('.alert').length).toBeGreaterThan(0);
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Be explicit about the length. Also, check the actual error message.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            it('should throw an error if submitted with blank fields', function (done) {
+                var mockMouseHandler = { preventDefault: function () {} };
+                view.formView.submit(mockMouseHandler);
+                setTimeout(function () {
+                    expect(view.$el.find('.alert').length).toBeGreaterThan(0);
+                    done();
+                }, 100);
+            });
+
+            it('should throw no errors if ', function (done) {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: This doesn't correspond with what you're testing.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                model = Course.findOrCreate(data, {parse: true});
+                view = new CourseCreateEditView({ model: model, editing: true }).render();
+            });
+
+            it('should display course details in form fields', function () {
+                var type_label = view.$el.find(".course-types input[type=radio]:checked").siblings('.courseTypeTitle');
+
+                expect(view.$el.find('.course-id').text()).toEqual(model.get('id'));
+                expect(view.$el.find('.course-name').val()).toEqual(model.get('name'));
+                expect(type_label.text()).toEqual(_s.capitalize(model.get('type')));
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: You could also just check the radio value.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 4 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: @clintonb Unrelated.. I believe this test is incomplete and was overlooked, so I'll remove it. To address your question, I tried calling jQuery's .click() function and it wasn't called. All research I did regarding the matter seemed to use workarounds to make the click functionality work. This was the simplest.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                        }
+                    ]
+                };
+
+            beforeEach(function () {
+                model = Course.findOrCreate(data, {parse: true});
+                view = new CourseCreateEditView({ model: model, editing: true }).render();
+            });
+
+            it('should display course details in form fields', function () {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: What about the seats?

File: ecommerce/static/js/views/course_create_edit_view.js [hunk 9]
```diff
@@ -4,7 +4,8 @@ define([
         'backbone.super',
         'underscore',
         'views/course_form_view',
-        'text!templates/course_create_edit.html'
+        'text!templates/course_create_edit.html',
+        'jquery-ui'
```
Comment (clintonb) [inline ecommerce/static/js/views/course_create_edit_view.js]: Remove.

File: ecommerce/static/templates/_course_type_radio_field.html [hunk 10]
```diff
@@ -4,7 +4,7 @@
                aria-describedby="courseType<%= type %>HelpBlock"
                <% if(disabled) { print('disabled="disabled"'); } %>
                <% if(checked) { print('checked'); } %>>
-        <%= displayName %>
+        <div class="courseTypeTitle" id="courseType<%= type %>Title"><%= displayName %></div>
```
Comment (clintonb) [inline ecommerce/static/templates/_course_type_radio_field.html]: `<span class="course-type-display-name"><%= displayName %></span>`

File: ecommerce/static/templates/course_form.html [hunk 11]
```diff
@@ -8,7 +8,7 @@
 
         <div class="course-id"><%= id %></div>
         <% } else { %>
-        <input type="text" class="form-control" name="id" placeholder="e.g. edX/DemoX/Demo_Course">
+        <input type="text" class="form-control course-id" name="id" placeholder="e.g. edX/DemoX/Demo_Course">
```
Comment (clintonb) [inline ecommerce/static/templates/course_form.html]: Remove.

File: ecommerce/static/templates/course_form.html [hunk 12]
```diff
@@ -17,7 +17,7 @@
     <!-- TODO: Hide this field and get the data from an API. -->
     <div class="form-group">
         <label class="hd-4" for="name"><%= gettext('Course Name') %></label>
-        <input type="text" class="form-control" name="name" placeholder="e.g. edX Demonstration Course">
+        <input type="text" class="form-control course-name" name="name" placeholder="e.g. edX Demonstration Course">
```
Comment (clintonb) [inline ecommerce/static/templates/course_form.html]: Remove.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                var type_label = view.$el.find(".course-types input[type=radio]:checked").siblings('.courseTypeTitle');
+
+                expect(view.$el.find('.course-id').text()).toEqual(model.get('id'));
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: `[name=id]`

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+                expect(view.$el.find('.course-name').val()).toEqual(model.get('name'));
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: `[name=name]`

File: bower.json (hunk 1 above)
Comment (clintonb) [inline bower.json]: Bootstrap: http://getbootstrap.com/javascript/#tooltips

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 6 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: @clintonb Like I said, it's useless and will be removed.

File: ecommerce/static/js/views/course_create_edit_view.js (hunk 9 above)
Comment (Nickersoft) [inline ecommerce/static/js/views/course_create_edit_view.js]: @clintonb This view requires jquery-ui to use the .tooltip() function

Comment (clintonb) [inline ecommerce/static/js/views/course_create_edit_view.js]: I wrote it. It does not.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 4 above)
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: https://api.jquery.com/category/events/event-object/

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
+define([
+        'jquery',
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Unless you opt to use `$.Event`, you don't need this requirement.

File: bower.json (hunk 1 above)
Comment (Nickersoft) [inline bower.json]: @clintonb Bootstrap JS wasn't included in this file, and it had been throwing an error. I was not aware tooltip() was a Bootstrap thing as well. I'll make the change.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 3 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

File: ecommerce/static/js/views/course_create_edit_view.js (hunk 9 above)
Comment (Nickersoft) [inline ecommerce/static/js/views/course_create_edit_view.js]: @clintonb It requires Bootstrap, which was not included in this file. Without the jquery-ui import, it was saying that tooltip() is not a function. I'll add the Bootstrap import.

File: bower.json (hunk 1 above)
Comment (clintonb) [inline bower.json]: Bootstrap is included. See `bootstrap-sass`.

File: ecommerce/static/templates/course_form.html (hunk 11 above)
Comment (Nickersoft) [inline ecommerce/static/templates/course_form.html]: Fixed

File: ecommerce/static/templates/course_form.html (hunk 12 above)
Comment (Nickersoft) [inline ecommerce/static/templates/course_form.html]: Fixed

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 5 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

File: bower.json (hunk 1 above)
Comment (Nickersoft) [inline bower.json]: @clintonb Correction: Bootstrap was not included in this RequireJS `define` statement.

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 2 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 5 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 13 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_create_edit_view_spec.js (hunk 14 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_edit_view_spec.js]: Fixed

**Comment by Nickersoft** [comment]
@clintonb @rlucioni @jimabramson Updated

File: ecommerce/static/js/test/specs/course_create_view_spec.js [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            beforeEach(function () {
+                model = new Course();
+                view = new CourseCreateEditView({ model: model, editing: false }).render();
+            });
+
+            it('should throw an error if submitted with blank fields', function (done) {
+                var mockMouseHandler = { preventDefault: function () {} },
+                    alertView = new AlertView({
+                        level: 'danger',
+                        title: gettext('Error!'),
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: No need for `gettext`.

File: ecommerce/static/js/test/specs/course_create_view_spec.js [hunk 17]
```diff
@@ -0,0 +1,43 @@
+define([
+        'jquery',
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Not used.

File: ecommerce/static/js/test/specs/course_create_view_spec.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        'underscore.string',
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Not used.

File: ecommerce/static/js/test/specs/course_create_view_spec.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+            it('should throw an error if submitted with blank fields', function (done) {
+                var mockMouseHandler = { preventDefault: function () {} },
+                    alertView = new AlertView({
+                        level: 'danger',
+                        title: gettext('Error!'),
+                        message: gettext('Please complete all required fields.')
+                    });
+
+                view.formView.submit(mockMouseHandler);
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: `view.formView.submit($.Event('click'));` definitely works. I just tried it on my own machine. Use it and get rid of `mockMouseHandler`.

File: ecommerce/static/js/test/specs/course_create_view_spec.js [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+                setTimeout(function () {
+                    expect(view.$el.find('.alert').length).toBe(1);
+                    expect(view.$el.find('.alert').html()).toBe(alertView.render().$el.html());
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Is this really a good test? What if `AlertView` doesn't render anything? Check the _actual content_.

File: ecommerce/static/js/test/specs/course_edit_view_spec.js [hunk 21]
```diff
@@ -0,0 +1,130 @@
+define([
+        'jquery',
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_edit_view_spec.js]: Unused. Remove.

File: ecommerce/static/js/test/specs/course_edit_view_spec.js [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+            it('should display correct course seats given course type', function () {
+                var honorView = new HonorCourseSeatFormFieldView({ model: model.getOrCreateSeat('honor') }),
+                    verifiedView = new VerifiedCourseSeatFormFieldView({ model: model.getOrCreateSeat('verified') }),
+                    honorElement = view.$el.find('.row.course-seat.honor'),
+                    verifiedElement = view.$el.find('.row.course-seat.verified');
+
+                expect(honorElement.length).toBe(1);
+                expect(verifiedElement.length).toBe(1);
+
+                expect(honorElement.html()).toBe(honorView.render().$el.html());
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/course_edit_view_spec.js]: Check the content. Until you have actually checked the rendered content of a view, you cannot use it in validation. Trust, but verify.

File: ecommerce/static/js/test/specs/course_create_view_spec.js (hunk 19 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: @clintonb I believe I misunderstood your recommendation. I was using $.Event along with .trigger() (as used in the jQuery docs) to mimic a click event. I agree, this should work. Sorry for the confusion.

File: ecommerce/static/js/test/specs/course_create_view_spec.js (hunk 16 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_create_view_spec.js (hunk 18 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_create_view_spec.js (hunk 20 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_create_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_edit_view_spec.js (hunk 21 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_edit_view_spec.js]: Fixed

File: ecommerce/static/js/test/specs/course_edit_view_spec.js (hunk 22 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/course_edit_view_spec.js]: Fixed

**Comment by Nickersoft** [comment]
@clintonb @rlucioni @jimabramson Updated

File: Makefile [hunk 23]
```diff
@@ -55,6 +55,9 @@ test_python: clean
 	--settings=ecommerce.settings.test --with-ignore-docstrings --logging-level=DEBUG
 	coverage report
 
+test_js:
```
Comment (clintonb) [inline Makefile]: Unnecessary. See `validate_js`.

File: ecommerce/settings/local.py [hunk 24]
```diff
@@ -85,13 +85,13 @@
 
 # AUTHENTICATION
 # Set these to the correct values for your OAuth2/OpenID Connect provider (e.g., devstack)
-SOCIAL_AUTH_EDX_OIDC_KEY = 'replace-me'
```
Comment (clintonb) [inline ecommerce/settings/local.py:88]: Revert all of these changes.

File: ecommerce/static/js/utils/utils.js [hunk 25]
```diff
@@ -48,7 +48,7 @@ define([
              */
             stripTimezone: function (datetime) {
                 if (datetime) {
-                    datetime = moment.utc(datetime).format('YYYY-MM-DDTHH:mm:ss');
+                    datetime = moment.utc(new Date(Date.parse(datetime))).format('YYYY-MM-DDTHH:mm:ss');
```
Comment (clintonb) [inline ecommerce/static/js/utils/utils.js]: Why was this changed? The whole purpose of `moment` is to aid in parsing?

File: ecommerce/static/js/test/specs/views/course_edit_view_spec.js [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+
+            it('should display correct course seats given course type', function () {
+                var honorElement = view.$el.find('.row.course-seat.honor'),
+                    verifiedElement = view.$el.find('.row.course-seat.verified'),
+                    formattedDate = new Date('2020-01-01T0:00:00Z');
+
+                expect(honorElement.length).toBe(1);
+                expect(verifiedElement.length).toBe(1);
+
+                expect(honorElement.find('.seat-price').text()).toBe('$0.00');
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/course_edit_view_spec.js:118]: Separate the honor and verified assertions. It's confusing to see the two price checks directly beside one another.

File: ecommerce/settings/local.py (hunk 24 above)
Comment (Nickersoft) [inline ecommerce/settings/local.py:88]: @clintonb This file was accidentally committed.. will do.

File: Makefile (hunk 23 above)
Comment (Nickersoft) [inline Makefile]: @clintonb This seems rather inconsistent seeing the Makefile help displays commands for `test_python` and not `validate_python`, yet there should be a `validate_js` command without a `test_js` command. Shouldn't the Makefile help either just include `validate_python` and `validate_js` or just include `test_python` and `test_js`?

File: ecommerce/static/js/utils/utils.js (hunk 25 above)
Comment (Nickersoft) [inline ecommerce/static/js/utils/utils.js]: @clintonb Due to depreciation warnings in the console 
https://github.com/moment/moment/issues/1407

File: ecommerce/static/js/test/specs/views/course_edit_view_spec.js (hunk 26 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/views/course_edit_view_spec.js:118]: Fixed

File: ecommerce/static/js/utils/utils.js (hunk 25 above)
Comment (clintonb) [inline ecommerce/static/js/utils/utils.js]: That's why we are using `moment.utc`. There are currently no console warnings in stage or prod. If you are seeing warnings, you should check your changes.

File: Makefile (hunk 23 above)
Comment (clintonb) [inline Makefile]: Everything will be changed to `validate` shortly.

File: ecommerce/static/js/test/specs/views/course_edit_view_spec.js [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+                var honorElement = view.$el.find('.row.course-seat.honor'),
+                    verifiedElement = view.$el.find('.row.course-seat.verified'),
+                    formattedDate = new Date('2020-01-01T0:00:00Z');
+
+                expect(honorElement.length).toBe(1);
+                expect(honorElement.find('.seat-price').text()).toBe('$0.00');
+
+                expect(verifiedElement.length).toBe(1);
+                expect(verifiedElement.find('[name=price]').val()).toBe('15.00');
+                expect(verifiedElement.find('[name=expires]').val()).toBe(moment(formattedDate).format());
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/course_edit_view_spec.js]: You don't really need `moment` here. Just use the string.

File: ecommerce/static/js/utils/utils.js (hunk 25 above)
Comment (Nickersoft) [inline ecommerce/static/js/utils/utils.js]: @clintonb The warning I'm receiving is explicitly for this line. I had not previously modified this file. I'm using version 2.10.6 when the Bower version is 2.10.3.. is there the possibility a recent change causes depreciation warnings to appear more often?

Comment (clintonb) [inline ecommerce/static/js/utils/utils.js]: You should not see any warnings here. The value of `datetime` is always in ISO format and the UTC timezone. Please revert this change.

Comment (Nickersoft) [inline ecommerce/static/js/utils/utils.js]: Fixed

File: ecommerce/static/js/test/specs/views/course_edit_view_spec.js (hunk 27 above)
Comment (Nickersoft) [inline ecommerce/static/js/test/specs/views/course_edit_view_spec.js]: Fixed

**Comment by Nickersoft** [comment]
@clintonb @jimabramson @rlucioni Updated

File: Makefile (hunk 23 above)
Comment (clintonb) [inline Makefile]: Remove.

File: ecommerce/static/js/test/specs/views/course_create_view_spec.js [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+
+            beforeEach(function () {
+                model = new Course();
+                view = new CourseCreateEditView({ model: model, editing: false }).render();
+            });
+
+            it('should throw an error if submitted with blank fields', function (done) {
+                var errorHTML = '<strong>Error!</strong> Please complete all required fields.';
+                view.formView.submit($.Event('click'));
+                setTimeout(function () {
```
Comment (clintonb) [inline ecommerce/static/js/test/specs/views/course_create_view_spec.js]: You don't need a timeout here since `submit` is synchronous.

**Comment by clintonb** [comment]
:+1: (once timeout addressed)

**Comment by Nickersoft** [comment]
@jimabramson @rlucioni ?

**Comment by clintonb** [comment]
@Nickersoft make sure you squash your commits.

File: Makefile (hunk 23 above)
Comment (rlucioni) [inline Makefile]: @Nickersoft make sure you address this. This is addressed in https://github.com/edx/ecommerce/pull/273.

**Comment by Nickersoft** [comment]
@clintonb Will do

**Comment by rlucioni** [comment]
@Nickersoft it looks like something went wrong with your rebase. This PR now includes a bunch of irrelevant changes from earlier in the repo's history.

**Comment by Nickersoft** [comment]
I believe everything has been resolved @clintonb and @rlucioni

**Comment by rlucioni** [comment]
Nice @Nickersoft. :+1:

**Comment by Nickersoft** [comment]
Will merge after tests pass

**Comment by jimabramson** [comment]
:+1:

**Comment by clintonb** [comment]
@Nickersoft I rebased your branch on master. If you pull, and rebase again, coverage should inch a little higher. Once that's done, merge away.
